src/Security/Authorization/Policy/src/AuthorizationMiddleware.cs

```diff
@@ -47,20 +47,31 @@ public async Task Invoke(HttpContext context)
         }
 
         var endpoint = context.GetEndpoint();
-
         if (endpoint != null)
         {
             // EndpointRoutingMiddleware uses this flag to check if the Authorization middleware processed auth metadata on the endpoint.
             // The Authorization middleware can only make this claim if it observes an actual endpoint.
             context.Items[AuthorizationMiddlewareInvokedWithEndpointKey] = AuthorizationMiddlewareWithEndpointInvokedValue;
         }
 
-        // IMPORTANT: Changes to authorization logic should be mirrored in MVC's AuthorizeFilter
-        var authorizeData = endpoint?.Metadata.GetOrderedMetadata<IAuthorizeData>() ?? Array.Empty<IAuthorizeData>();
+        // Use the computed policy for this endpoint if we can
+        var computedPolicy = endpoint?.Metadata.GetMetadata<AuthorizationPolicyCache>();
+        var policy = computedPolicy?.Policy;
+        if (policy == null)
```

Comment:
This is probably very rare but this would be a breaking change for scenarios when a user have its own `IAuthorizationPolicyProvider` and the policy is dynamic.  https://github.com/dotnet/aspnetcore/blob/5825417f345bf0ffd448c5f5542be83558105317/src/Security/Authorization/Core/src/AuthorizationPolicy.cs#L162

Reply:
@DamianEdwards @davidfowl should we do the same behavior as MVC and only cache if the policy provider is the default one to be safe?

Reply:
Also great catch @Kahbazi !

Reply:
Yes we should only cache the default to reduce the chance of breaking people doing crazy stuff like injecting the IHttpContextAccessor 😄